core/common/test/DateTimePeriodTest.kt

```diff
@@ -29,8 +29,43 @@ class DateTimePeriodTest {
         assertEquals("-P1DT1H", DateTimePeriod(days = -1, hours = -1).toString())
         assertEquals("-P1M", DateTimePeriod(months = -1).toString())
 
-        assertEquals("P-1Y-2M-3DT-4H-5M0.500000000S",
+        assertEquals("-P1Y2M3DT4H4M59.500000000S",
                 DateTimePeriod(years = -1, months = -2, days = -3, hours = -4, minutes = -5, seconds = 0, nanoseconds = 500_000_000).toString())
+
+        assertEquals("PT277H46M39.999999999S", DateTimePeriod(nanoseconds = 999_999_999_999_999L).toString())
+        assertEquals("PT0.999999999S", DateTimePeriod(seconds = 1, nanoseconds = -1L).toString())
+        assertEquals("-PT0.000000001S", DateTimePeriod(nanoseconds = -1L).toString())
+        assertEquals("P1DT-0.000000001S", DateTimePeriod(days = 1, nanoseconds = -1L).toString())
+        assertEquals("-PT0.999999999S", DateTimePeriod(seconds = -1, nanoseconds = 1L).toString())
+        assertEquals("P1DT-0.999999999S", DateTimePeriod(days = 1, seconds = -1, nanoseconds = 1L).toString())
+    }
+
+    @Test
+    fun parseIsoString() {
+        assertEquals(DateTimePeriod(years = 1), DateTimePeriod.parse("P1Y"))
+        assertEquals(DatePeriod(years = 1, months = 1), DateTimePeriod.parse("P1Y1M"))
+        assertEquals(DateTimePeriod(months = 11), DateTimePeriod.parse("P11M"))
+        assertEquals(DateTimePeriod(months = 14), DateTimePeriod.parse("P14M"))
+        assertEquals(DateTimePeriod(months = 10, days = 5), DateTimePeriod.parse("P10M5D"))
+        assertEquals(DateTimePeriod(years = 1, days = 40), DateTimePeriod.parse("P1Y40D"))
+
+        assertEquals(DateTimePeriod(hours = 1), DateTimePeriod.parse("PT1H"))
+        assertEquals(DateTimePeriod(), DateTimePeriod.parse("P0D"))
+        assertEquals(DatePeriod(), DateTimePeriod.parse("P0D"))
+
+        assertEquals(DateTimePeriod(days = 1, hours = -1), DateTimePeriod.parse("P1DT-1H"))
+        assertEquals(DateTimePeriod(days = -1, hours = -1), DateTimePeriod.parse("-P1DT1H"))
+        assertEquals(DateTimePeriod(months = -1), DateTimePeriod.parse("-P1M"))
+
+        assertEquals(DateTimePeriod(years = -1, months = -2, days = -3, hours = -4, minutes = -5, seconds = 0, nanoseconds = 500_000_000),
+            DateTimePeriod.parse("P-1Y-2M-3DT-4H-5M0.500000000S"))
+
+        assertEquals(DateTimePeriod(nanoseconds = 999_999_999_999_999L), DateTimePeriod.parse("PT277H46M39.999999999S"))
+        assertEquals(DateTimePeriod(seconds = 1, nanoseconds = -1L), DateTimePeriod.parse("PT0.999999999S"))
+        assertEquals(DateTimePeriod(nanoseconds = -1L), DateTimePeriod.parse("-PT0.000000001S"))
+        assertEquals(DateTimePeriod(days = 1, nanoseconds = -1L), DateTimePeriod.parse("P1DT-0.000000001S"))
+        assertEquals(DateTimePeriod(seconds = -1, nanoseconds = 1L), DateTimePeriod.parse("-PT0.999999999S"))
+        assertEquals(DateTimePeriod(days = 1, seconds = -1, nanoseconds = 1L), DateTimePeriod.parse("P1DT-0.999999999S"))
```

Comment:
A couple of negative tests won't hurt. For example that DatePeriod.parse fails on non-zero time components.

Reply:
Added some.